src/engine/qcommon/q_math.cpp

```diff
@@ -1082,7 +1082,7 @@ void PerpendicularVector( vec3_t dst, const vec3_t src )
 	/*
 	 * * normalize the result
 	 */
-	VectorNormalize( dst );
+	VectorNormalizeFast( dst );
```

Comment:
I definitely don't want to see the inaccurate version used for basic math routines. It's OK to use the sloppy version for graphics-specific stuff

Reply:
Why is it inaccurate? They both use `Q_sqrt()` (and then they both use `_mm_rsqrt_ps()`).

```c++
// returns vector length
inline vec_t VectorNormalize( vec3_t v )
{
	vec_t length = DotProduct( v, v );

	if ( length != 0.0f )
	{
		vec_t ilength = Q_rsqrt( length );
		/* sqrtf(length) = length * (1 / sqrtf(length)) */
		length *= ilength;
		VectorScale( v, ilength, v );
	}

	return length;
}

// does NOT return vector length, uses rsqrt approximation
// fast vector normalize routine that does not check to make sure
// that length != 0, nor does it return length
inline void VectorNormalizeFast( vec3_t v )
{
	vec_t ilength = Q_rsqrt( DotProduct( v, v ) );

	VectorScale( v, ilength, v );
}
```

The comment before `VectorNormalizeFast()` saying `uses rsqrt approximation` is misleading, they both uses rsqrt. Maybe long time ago in the past they didn't, but they do.

The difference is that `VectorNormalize()` tests for the length not being zero, skips an useless scale if it is, and returns the length.

In the past I did the contrary, replacing `VectorNormalizeFast()` with `VectorNormalize()` in hope the extra test would actually make it faster by skipping the scale, it was slower because of the test breaking the vectorization.

The two functions should produce the same results as they use the same operations, one of them is just slower.

Reply:
OK I commented on an irrelevant diff here, but I stand by the words themselves. The old version of Q_rsqrt was pretty accurate, having only the least significant 1-2 binary places wrong, so it was good enough to use more or less anywhere. Now if we are going to make it a rough approximation, we shouldn't be using it all the time.